### ISSUE TITLE
Potential oversight or incompatibility in `SFTTrainer`: `assistant_only_loss=True` silently ignored when `use_liger_kernel=True`

### ISSUE THREAD

**Issue by andylvua** [huggingface/trl#3781]
## Summary

When `assistant_only_loss=True` is used in combination with `use_liger_kernel=True` in `trl.SFTConfig`, the `assistant_masks` are silently discarded during dataset preparation. This causes the model to compute loss over the entire sequence, contrary to the user’s configuration.

There is no mention of this incompatibility in the official documentation, and no warning is raised at runtime.

---

## Context

According to the documentation:

* [`[assistant_only_loss]`](https://huggingface.co/docs/trl/v0.19.1/en/sft_trainer#trl.SFTConfig.assistant_only_loss):

  > Whether to compute loss only on the assistant part of the sequence. If set to True, loss is computed only on the assistant responses...

* [`[use_liger_kernel]`](https://huggingface.co/docs/transformers/v4.53.1/en/main_classes/trainer#transformers.TrainingArguments.use_liger_kernel):

  > Whether to enable Liger Kernel for LLM model training... supports flash attention, FSDP, DeepSpeed...

* The [SFT guide](https://huggingface.co/docs/trl/sft_trainer#liger-kernel-increase-20-throughput-and-reduce-60-memory-for-multi-gpu-training) also does not mention any limitation or restriction regarding `assistant_only_loss`.

---

## Observed Behavior

In the source code of `trl.trainer.sft_trainer.SFTTrainer`, the assistant mask is correctly created if `assistant_only_loss=True`:

https://github.com/huggingface/trl/blob/accf7383a33c618a2edc7205107120b5f32e28a3/trl/trainer/sft_trainer.py#L756-L773

However, later in the same file, the assistant mask is silently excluded if Liger kernel is enabled:

https://github.com/huggingface/trl/blob/accf7383a33c618a2edc7205107120b5f32e28a3/trl/trainer/sft_trainer.py#L801-L803

This discards `assistant_masks` silently, even though they were previously computed and are required for `assistant_only_loss`. Consequently, this breaks the masking logic applied in the default data collator:

https://github.com/huggingface/trl/blob/accf7383a33c618a2edc7205107120b5f32e28a3/trl/trainer/sft_trainer.py#L249-L253

To the best of our understanding, there is no inherent incompatibility between Liger kernel and assistant masking. The assistant masks are utilized exclusively within the data collator, where they are applied to selectively mask label tokens by assigning a value of `-100`. This mechanism is equivalent in principle to conventional token-level masking strategies, such as the use of attention masks for padding, and should not interfere with the computational or memory optimizations introduced by the Liger kernel.

Therefore, the current behavior appears to be an implementation oversight, rather than a deliberate constraint.

---

## Proposed Resolution

To ensure correct behavior, `assistant_masks` should be retained when `use_liger_kernel=True`. The following patch resolves the issue:

```diff
# trl/trainer/sft_trainer.py
if args.use_liger_kernel:
-    dataset = dataset.select_columns({"input_ids", "position_ids"}.intersection(dataset.column_names))
+    dataset = dataset.select_columns({"input_ids", "position_ids", "assistant_masks"}.intersection(dataset.column_names))
```

This change has been tested and confirmed to work correctly. Assistant masks are passed to the collator, which applies them appropriately, resulting in correct loss computation.

Therefore:
* If `assistant_only_loss=True` is intended to be supported with Liger kernel, the above patch or similar one should be applied.
* If there is an architectural reason why assistant masks should not be used with Liger kernel, this should be:

  * Explicitly stated in the documentation.
  * Enforced by a runtime error or warning when both options are enabled.

---

**Thanks for the excellent work on the `trl` library!**

### System Info

- Platform: Linux-6.8.0-64-generic-x86_64-with-glibc2.35
- Python version: 3.10.18
- TRL version: 0.19.1
- PyTorch version: 2.7.1
- accelerator(s): 3xNVIDIA GeForce RTX 4090
- Transformers version: 4.53.3
- Datasets version: 4.0.0
- HF Hub version: 0.34.1
- bitsandbytes version: 0.46.1
- Diffusers version: 0.34.0
- Liger-Kernel version: 0.6.0
- PEFT version: 0.16.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Thank!! Do you want to open a PR to solve this?

**Comment by andylvua** [comment]
Sure! I’d be happy to open a PR to address this.
